### ISSUE TITLE
[PT2] Graph break in forward pre-hook skips compiling forward for `nn.Transformer`

### ISSUE THREAD

**Issue by awgu** [pytorch/pytorch#111966]
If there is a graph break in the forward pre-hook (`nn.Module.register_forward_pre_hook()`), then the `nn.Transformer.forward()` gets skipped from being compiled.
```
from typing import Any

import torch
import torch.distributed as dist
import torch.nn as nn
import torch._dynamo

def forward_pre_hook_with_graph_break(module: nn.Module, inp: Any):
    torch._dynamo.graph_break()
    return inp

def forward_pre_hook(module: nn.Module, inp: Any):
    return inp

def run_transformer_1d():
    torch.manual_seed(42)
    model = nn.Transformer().cuda()
    model.register_forward_pre_hook(forward_pre_hook_with_graph_break)
    # model.register_forward_pre_hook(forward_pre_hook)
    model = torch.compile(model)
    src = torch.randn((20, 512, 512), device="cuda")
    tgt = torch.randn((10, 512, 512), device="cuda")
    for _ in range(3):
        out = model(src, tgt).sum()
        out.backward()

if __name__ == "__main__":
    torch.cuda.set_device("cuda:0")
    run_transformer_1d()
```

`TORCH_LOGS="+dynamo"` with graph break in the hook (no compiled function!): P863664749
`TORCH_LOGS="+dynamo"` without graph break in the hook (compiled function!): P863667227

@ezyang said that this is "~expected" and to cc @yanboliang. FWIW, I am less concerned with compiling `nn.Transformer` specifically. This is related to FSDP, where I want to `torch._dynamo.disable` the module hooks but still compile the module forward. If there is not a good general solution, then we can revisit the FSDP issue specifically (i.e. we do not need to one-off fix this for `nn.Transformer` since it is only being used as the example model here).

Version: main

**Comment by yanboliang** [comment]
This is sort of expected behavior, but I don't think it's correct. Let me explain the reason behind of the behavior:
* Dynamo is tracing and inlining ```nn.Module.__call__/_wrapped_call_impl```;
* Then it recursively inlines ```forward_pre_hook``` if there is;
* Graph breaks inside of ```forward_pre_hook``` caused the compilation of ```nn.Module.__call__/_wrapped_call_impl``` failed and fallback to eager.
* Dynamo recursively evaluates the frame of each function inside of ```nn.Module.__call__/_wrapped_call_impl```, e.g., ```forward```, but it doesn't pass [this line](https://github.com/pytorch/pytorch/blob/4f42edfb6e5b703eec2a14b8933090646702c5a2/torch/_dynamo/eval_frame.py#L531) and skips this frame.
  * The reason behind this rule is, we skip all recursively called frame if the upper level frame is skipped, which is true in most of the cases. Otherwise, we would recursively trace many low level frames, which is not efficient and may trigger a lot of features that Dynamo doesn't supported.

This is a pretty reasonable case to not skip the underlying frame, I think we can support this. The straightforward idea is to check if the function is ```torch.nn.Module.forward```. I'll send a PR for this.

**Comment by ezyang** [comment]
I don't agree with the diagnosis here.

The part which I do not agree with is the sub-bullet here:

> The reason behind this rule is, we skip all recursively called frame if the upper level frame is skipped, which is true in most of the cases. Otherwise, we would recursively trace many low level frames, which is not efficient and may trigger a lot of features that Dynamo doesn't supported.

I don't think we "skip all recursively called frame if the upper level frame is skipped". If you look at the current skipfiles.check_verbose, there's nothing in it that suggests that we will start skipping things if an outer frame was skipped (which would require some sort of global context tracking if we had skipped an outer frame)

```
def check_verbose(obj, allow_torch=False):
    if isinstance(
        obj, (UserFunctionVariable, UserMethodVariable, NestedUserFunctionVariable)
    ):
        filename = obj.get_filename()
        obj = obj.get_code()
    elif isinstance(obj, types.CodeType):
        filename = obj.co_filename
    elif isinstance(obj, (types.FunctionType, types.MethodType)):
        filename = getfile(obj)
        obj = obj.__code__
    else:
        filename = getfile(obj)
    if obj in get_func_inlinelist():
        return SkipResult(
            False,
            "inlined according skipfiles.FUNC_INLINELIST",
        )
    return check_file(filename, allow_torch)
```

Additionally, when I modify the example program so that I am calling a user-defined Transformer, this works fine:

```
from typing import Any

import torch
import torch.distributed as dist
import torch.nn as nn
import torch._dynamo

def forward_pre_hook_with_graph_break(module: nn.Module, inp: Any):
    torch._dynamo.graph_break()
    return inp

def forward_pre_hook(module: nn.Module, inp: Any):
    return inp

class MyTransformer(nn.Transformer):
    def forward(self, *args, **kwargs):
        return super().forward(*args, **kwargs)

def run_transformer_1d():
    torch.manual_seed(42)
    model = MyTransformer().cuda()
    model.register_forward_pre_hook(forward_pre_hook_with_graph_break)
    # model.register_forward_pre_hook(forward_pre_hook)
    model = torch.compile(model)
    src = torch.randn((20, 512, 512), device="cuda")
    tgt = torch.randn((10, 512, 512), device="cuda")
    for _ in range(3):
        out = model(src, tgt).sum()
        out.backward()

if __name__ == "__main__":
    torch.cuda.set_device("cuda:0")
    run_transformer_1d()
```

In other words, the root cause is that by default, we are unwilling to torch.compile NN modules that are directly defined in PyTorch library, unless there is some user code that calls it. But we should then change the policy on a case-by-case basis. Do you want to compile nn.Linear if it's called all by itself? Probably not. @awgu, would you be happy if it were *only* Transformer that got special cased in this way? You implied it doesn't matter. I just want to point out that for any realistic use case of FSDP, what will be wrapped is a user module, and we will trace into it anyway.

**Comment by awgu** [comment]
@ezyang Thanks for the context! This makes sense.

I think we do not need to special case `nn.Transformer` until we have an explicit ask for it (as most people using FSDP are not using `nn.Transformer` but rather some user-defined transformer like you mentioned).

For user-defined modules, I can actually get per-parameter-sharding FSDP to compile with graph breaks on FSDP logic (like existing FSDP) using `FSDPManagedNNModuleVariable` (though `torch.distributed._composble.checkpoint` does not work), so I think that the issue mentioned here is not too important anymore. Sorry everyone for the confusion!

**Comment by anijain2305** [comment]
Marking low prio as this is expected behavior and User defined nn modules work well.
